File: SecurityAngelIOS/Services/FamilyRepository.swift
import Foundation
import FirebaseFirestore

/// Family CRUD + member management. Mirrors Android's `FamilyRepository`.
final class FamilyRepository {

    private var db: Firestore { Firestore.firestore() }
    private var families: CollectionReference { db.collection("families") }
    private var users: CollectionReference { db.collection("users") }

    enum FamilyError: LocalizedError {
        case userNotFound
        case alreadyInFamily
        case familyNotFound

        var errorDescription: String? {
            switch self {
            case .userNotFound:     return "User with this email not found."
            case .alreadyInFamily:  return "User already belongs to a family."
            case .familyNotFound:   return "Family not found."
            }
        }
    }

    // MARK: - Create

    func createFamily(admin: SecurityUser, name: String) async throws -> String {
        let ref = families.document()
        let family = SecurityFamily(
            id: ref.documentID,
            adminId: admin.id,
            name: name,
            members: [admin.id]
        )
        let batch = db.batch()
        try batch.setData(from: family, forDocument: ref)
        batch.updateData(["familyId": ref.documentID], forDocument: users.document(admin.id))
        try await batch.commit()
        return ref.documentID
    }

    // MARK: - Add / remove members

    func addMember(familyId: String, byEmail email: String) async throws {
        let lookup = try await users.whereField("email", isEqualTo: email).limit(to: 1).getDocuments()
        guard let doc = lookup.documents.first else { throw FamilyError.userNotFound }
        if let existing = doc.get("familyId") as? String, !existing.isEmpty {
            throw FamilyError.alreadyInFamily
        }
        let memberId = doc.documentID
        let batch = db.batch()
        batch.updateData(
            ["members": FieldValue.arrayUnion([memberId])],
            forDocument: families.document(familyId)
        )
        batch.updateData(["familyId": familyId], forDocument: users.document(memberId))
        try await batch.commit()
    }

    func removeMember(familyId: String, memberId: String) async throws {
        let batch = db.batch()
        batch.updateData(
            ["members": FieldValue.arrayRemove([memberId])],
            forDocument: families.document(familyId)
        )
        batch.updateData(["familyId": FieldValue.delete()], forDocument: users.document(memberId))
        try await batch.commit()
    }

    // MARK: - Read

    func get(id: String) async throws -> SecurityFamily? {
        let snapshot = try await families.document(id).getDocument()
        guard snapshot.exists else { return nil }
        var family = try snapshot.data(as: SecurityFamily.self)
        family.id = snapshot.documentID
        return family
    }

    func observeMembers(familyId: String, onChange: @escaping ([SecurityUser]) -> Void) -> ListenerRegistration {
        users.whereField("familyId", isEqualTo: familyId).addSnapshotListener { snapshot, _ in
            let members = snapshot?.documents.compactMap { snap -> SecurityUser? in
                guard var u = try? snap.data(as: SecurityUser.self) else { return nil }
                u.id = snap.documentID
                return u
            } ?? []
            onChange(members)
        }
    }

    // MARK: - Join via invitation

    /// Atomically attach `userId` to `familyId` and remove the invitation.
    func joinFamily(userId: String, familyId: String, invitationEmail: String) async throws {
        let batch = db.batch()
        batch.updateData(["familyId": familyId], forDocument: users.document(userId))
        batch.updateData(
            ["members": FieldValue.arrayUnion([userId])],
            forDocument: families.document(familyId)
        )
        batch.deleteDocument(db.collection("invitations").document(invitationEmail))
        try await batch.commit()
    }
}

File: SecurityAngel/app/src/main/java/com/example/securityangel/data/repo/FamilyRepository.kt
package com.example.securityangel.data.repo

import com.example.securityangel.data.models.Family
import com.example.securityangel.data.models.User
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FieldValue

object FamilyRepository {

    private val db = FirebaseFirestore.getInstance()
    private val usersRef = db.collection("users")
    private val familiesRef = db.collection("families")

    fun createFamily(admin: User, familyName: String, onSuccess: (String) -> Unit, onFailure: (String) -> Unit) {
        val newFamilyRef = familiesRef.document()
        val familyId = newFamilyRef.id

        val newFamily = Family(
            id = familyId,
            adminId = admin.id,
            name = familyName,
            members = listOf(admin.id)
        )

        db.runBatch { batch ->
            batch.set(newFamilyRef, newFamily)
            batch.update(usersRef.document(admin.id), "familyId", familyId)
        }.addOnSuccessListener {
            onSuccess(familyId)
        }.addOnFailureListener { e ->
            onFailure(e.message ?: "Error creating family")
        }
    }

    fun addMemberByEmail(familyId: String, email: String, onSuccess: () -> Unit, onFailure: (String) -> Unit) {

        usersRef.whereEqualTo("email", email).get()
            .addOnSuccessListener { snapshot ->
                if (snapshot.isEmpty) {
                    onFailure("User with this email not found")
                    return@addOnSuccessListener
                }

                val userDoc = snapshot.documents.first()
                val userId = userDoc.id

                if (userDoc.getString("familyId") != null) {
                    onFailure("User already belongs to a family")
                    return@addOnSuccessListener
                }

                db.runBatch { batch ->
                    batch.update(familiesRef.document(familyId), "members", FieldValue.arrayUnion(userId))
                    batch.update(usersRef.document(userId), "familyId", familyId)
                }.addOnSuccessListener {

                    onSuccess()
                }.addOnFailureListener { e ->

                    onFailure(e.message ?: "Failed to add member")
                }
            }
            .addOnFailureListener {
                onFailure("Error searching for user")
            }
    }

    fun getFamilyMembers(familyId: String, onResult: (List<User>) -> Unit) {
        usersRef.whereEqualTo("familyId", familyId).get()
            .addOnSuccessListener { snapshot ->
                val members = snapshot.toObjects(User::class.java)
                onResult(members)
            }
            .addOnFailureListener {
                onResult(emptyList())
            }
    }

    fun getFamilyData(familyId: String, onSuccess: (Family) -> Unit, onFailure: (String) -> Unit) {
        familiesRef.document(familyId).get()
            .addOnSuccessListener { document ->
                val family = document.toObject(Family::class.java)
                if (family != null) {
                    onSuccess(family)
                } else {
                    onFailure("Family not found")
                }
            }
            .addOnFailureListener { e ->
                onFailure(e.message ?: "Error fetching family")
            }
    }

    fun removeMember(familyId: String, memberId: String, onSuccess: () -> Unit, onFailure: (String) -> Unit) {
        db.runBatch { batch ->
            batch.update(familiesRef.document(familyId), "members", FieldValue.arrayRemove(memberId))

            batch.update(usersRef.document(memberId), "familyId", null)
        }.addOnSuccessListener {
            onSuccess()
        }.addOnFailureListener { e ->
            onFailure(e.message ?: "Failed to remove member")
        }
    }
}
